src/sentry/static/sentry/app/components/search/sources/index.jsx

```diff
@@ -47,26 +37,25 @@ class SearchSources extends React.Component {
     });
   }
 
-  render() {
+  renderSources(sources, results, idx) {
+    if (idx >= sources.length) {
+      return this.renderResults(results);
+    }
+    let Source = sources[idx];
     return (
-      <ApiSource {...this.props}>
-        {apiArgs => (
-          <FormSource {...this.props}>
-            {formFieldArgs => (
-              <RouteSource {...this.props}>
-                {routeArgs => (
-                  <CommandSource {...this.props}>
-                    {commandArgs =>
-                      this.renderResults(apiArgs, formFieldArgs, routeArgs, commandArgs)}
-                  </CommandSource>
-                )}
-              </RouteSource>
-            )}
-          </FormSource>
-        )}
-      </ApiSource>
+      <Source {...this.props}>
+        {args => {
+          results[idx] = args;
+          return this.renderSources(sources, results, idx + 1);
+        }}
+      </Source>
     );
   }
+
+  render() {
+    let {sources} = this.props;
+    return this.renderSources(sources, new Array(sources.length), 0);
```

Comment:
You don't need to allocate the array here, I think just pass an empty array, and we can just push args onto `results`.

Reply:
mutating at position is actually important because we don't control how children() is called and if we push and it's called multiple times we'll end up pushing duplicate results for everything down the chain :/

Reply:
Ah I would add a comment describing that then